Deprecate DisableLegacyLimitWrites & UseKvLimitsForNewOrder flags; remove code using certificatesPerName & newOrdersRL tables

jprenken opened the pull request:
Remove code using `certificatesPerName` & `newOrdersRL` tables.

Deprecate `DisableLegacyLimitWrites` & `UseKvLimitsForNewOrder` flags.

Remove legacy `ratelimit` package.

Delete these RA test cases:

- `TestAuthzFailedRateLimitingNewOrder` (rl: `FailedAuthorizationsPerDomainPerAccount`)
- `TestCheckCertificatesPerNameLimit` (rl: `CertificatesPerDomain`)
- `TestCheckExactCertificateLimit` (rl: `CertificatesPerFQDNSet`)
- `TestExactPublicSuffixCertLimit` (rl: `CertificatesPerDomain`)

Rate limits in NewOrder are now enforced by the WFE, starting here: https://github.com/letsencrypt/boulder/blob/5a9b4c4b18fd0aa670bc6332bdd59701ff7d6186/wfe2/wfe.go#L781

We collect a batch of transactions to check limits, check them all at once, go through and find which one(s) failed, and serve the failure with the Retry-After that's furthest in the future. All this code doesn't really need to be tested again; what needs to be tested is that we're returning the correct failure. That code is `NewOrderLimitTransactions`, and the `ratelimits` package's tests cover this.

The public suffix handling behavior is tested by `TestFQDNsToETLDsPlusOne`: https://github.com/letsencrypt/boulder/blob/5a9b4c4b18fd0aa670bc6332bdd59701ff7d6186/ratelimits/utilities_test.go#L9

Some other RA rate limit tests were deleted earlier, in #7869.

Part of #7671.

Johnbtc-bbc44nba commented on mocks/sa.go:
Okay i have found out, thank you so much for your wonderful help to mankind.....
Obenku johnbtc1

jprenken commented:
IN-10906

aarongable commented on cmd/boulder-ra/main.go:
```diff
@@ -31,11 +31,12 @@ import (
 
 type Config struct {
 	RA struct {
+		// RateLimitPoliciesFilename is deprecated.
```
In general we don't move struct fields to the top when we deprecate them -- we only do that in features.go because we frequently have many deprecated feature flags and it's nice to put them all in one spot. For a more generic deprecation like this, I'd leave the field in place so that things like `git blame` work better.

aarongable commented on sa/proto/sa.proto:
```diff
@@ -150,11 +146,6 @@ message Timestamps {
-  repeated string dnsNames = 2;
-}
-
 message CountByNames {
```
You should be able to delete this message too.

aarongable commented on sa/sa.go:
```diff
@@ -525,11 +490,11 @@ func (ssa *SQLStorageAuthority) AddCertificate(ctx context.Context, req *sapb.Ad
+	if fqdnTransactionErr != nil {
 		ssa.rateLimitWriteErrors.Inc()
-		ssa.log.AuditErrf("failed AddCertificate ratelimit update transaction: %v", rlTransactionErr)
+		ssa.log.AuditErrf("failed AddCertificate FQDN sets update transaction: %v", fqdnTransactionErr)
```
```suggestion
		ssa.log.AuditErrf("failed AddCertificate FQDN sets insert transaction: %v", fqdnTransactionErr)
```

aarongable commented on test/config-next/ra.json:
```diff
@@ -127,7 +126,6 @@
 		},
 		"features": {
 			"AsyncFinalize": true,
-			"UseKvLimitsForNewOrder": true,
```
Make sure this flag _is_ set in the corresponding //test/config/ra.json, to reflect the fact that it's currently set in prod. Same comment for SA and WFE below.

aarongable commented on test/config/ra.json:
```diff
@@ -1,6 +1,30 @@
+					"keyFile": "test/certs/ipki/wfe.boulder/key.pem"
+				}
+			},
+			"Defaults": "test/config-next/wfe2-ratelimit-defaults.yml",
+			"Overrides": "test/config-next/wfe2-ratelimit-overrides.yml"
```
Feels weird to be referencing config-next files from config. Probably best to move those files over to //test/config/, or at least put them somewhere other than //test/config-next/.

Same comment for wfe2.json below.

aarongable reviewed (commented):
Just a few small nits, and one high-level question: can the PR description grow a paragraph describing how the deleted RA test cases are covered by WFE kv-limit test cases?

jprenken commented:
> Just a few small nits, and one high-level question: can the PR description grow a paragraph describing how the deleted RA test cases are covered by WFE kv-limit test cases?

Addressed and added, thanks! I think there does need to be a replacement for `TestExactPublicSuffixCertLimit`, which I'll work up and add to this PR.

jsha commented on ra/ra.go:
```diff
@@ -1189,11 +1068,6 @@ func (ra *RegistrationAuthorityImpl) issueCertificateOuter(
 // errors from this function to the Subscriber, spends against these limit are
 // best effort.
 func (ra *RegistrationAuthorityImpl) countCertificateIssued(ctx context.Context, regId int64, orderDomains []string, isRenewal bool) {
-	if ra.limiter == nil || ra.txnBuilder == nil {
```
(which can happen as a followup)

beautifulentropy replied:
(also as a followup) If we aren't already we should "require" the relevant configuration fields.

jprenken replied:
Made #7951
